h2o-algos/src/main/java/hex/tree/isoforextended/ExtendedIsolationForest.java

```diff
@@ -0,0 +1,377 @@
+    }
+
+    @Override
+    protected Driver trainModelImpl() {
+        return new ExtendedIsolationForestDriver();
+    }
+
+    @Override
+    public ModelCategory[] can_build() {
+        return new ModelCategory[]{
+                ModelCategory.AnomalyDetection
+        };
+    }
+
+    @Override
+    protected ScoreKeeper.ProblemType getProblemType() {
+        return ScoreKeeper.ProblemType.anomaly_detection;
+    }
+
+    @Override
+    public boolean scoreZeroTrees() {
+        return false;
+    }
+
+    @Override
+    public boolean isSupervised() {
+        return false;
+    }
+
+    @Override
+    public boolean havePojo() {
+        return false;
+    }
+
+    @Override
+    public boolean haveMojo() {
+        return true;
+    }
+
+    @Override
+    protected void addCustomInfo(ExtendedIsolationForestModel.ExtendedIsolationForestOutput out) {
+        if (_iTrees != null) {
+            out.iTrees = _iTrees;
+        }
+    }
+
+    private class ExtendedIsolationForestDriver extends Driver {
+
+        @Override
+        protected ExtendedIsolationForestModel makeModel(Key<ExtendedIsolationForestModel> modelKey,
+                                                         ExtendedIsolationForestModel.ExtendedIsolationForestParameters parms) {
+            return new ExtendedIsolationForestModel(modelKey, parms, new ExtendedIsolationForestModel.ExtendedIsolationForestOutput(ExtendedIsolationForest.this));
+        }
+
+        @Override
+        protected boolean doOOBScoring() {
+            return true;
+        }
+
+        @Override
+        protected void scoreAndBuildTrees(boolean oob) {
+            IsolationTreeForkJoinTask [] iTreeTasks = new IsolationTreeForkJoinTask[_parms._ntrees];
+            for (int t = 0; t < _parms._ntrees; t++) {
+                iTreeTasks[t] = new IsolationTreeForkJoinTask();
+                H2O.submitTask(iTreeTasks[t]);
+            }
+            for (int t = 0; t < _parms._ntrees; t++) {
+                _iTrees[t] = iTreeTasks[t].getResult();
+            }
+            addCustomInfo(_model._output);
+        }
+
+        @Override
+        protected boolean buildNextKTrees() {
+            // this method is not used in the current implementation
+            // tree building is implemented in scoreAndBuildTrees method
+            return false;
+        }
+
+        @Override
+        protected void initializeModelSpecifics() {
+            _iTrees = new IsolationTree[_parms._ntrees];
+        }
+
+    }
+
+    /**
+     * IsolationTree class implements Algorithm 2 (iTree) and Algorithm 3 (pathLength) 
+     * from the Extended Isolation Forest paper.
+     * 
+     * The binary tree recursive implementation is replaced by array implementation.
+     * It is the most simple array implementation of binary tree.
+     */
+    public static class IsolationTree extends Iced<IsolationTree> {
+        private Node[] _nodes;
+
+        private Key<Frame> _frameKey;
```

Comment:
I think most of these could be final

Reply:
Will be done in https://github.com/h2oai/h2o-3/pull/4907